Add UI for release revisions

helfi92 opened the pull request:
Fixes https://github.com/mozilla-frontend-infra/balrog-ui/issues/64.
Relates to #104.

Note: I edited the title of https://github.com/mozilla-frontend-infra/balrog-ui/issues/104 to also virtualize the list of release revisions.

helfi92 commented:
![Screen Shot 2019-08-07 at 11 25 09 AM](https://user-images.githubusercontent.com/3766511/62635643-0d901700-b906-11e9-986c-f529254b1878.png)

mozbhearsum commented on src/App.jsx:
```diff
@@ -10,6 +10,10 @@ import theme from './theme';
 import Main from './Main';
 
 axios.interceptors.request.use(config => {
+  if (config.url.startsWith('http')) {
```
I assume this is to avoid sending credentials to Google and/or Auth0. Would it be better to check for a specific domain(s) instead of a protocol? Looking for just `http` feels like it's bound to break in the future. Or perhaps this is an argument for using Axios instances instead of a global? The latter might be out of scope for this PR.

helfi92 replied:
I think we never want to send the access token nor set `result.baseURL = BASE_URL;` when we are not sending a request to `BASE_URL`. I added a patch to add all of this logic inside a conditional:

```jsx
axios.interceptors.request.use(config => {
  const result = config;

  if (!config.url.startsWith('http')) {
    // set Authorization header
    // set baseURL
  }

  return result;
});
```

I think this will make it resilient to any future additions to that function. Let me know if you disagree. Happy to discuss.

mozbhearsum replied:
Ah, I think I get it now. Any Balrog backend URLs will only have a path component in them, whereas others will be a full URL.

helfi92 replied:
Yep.

mozbhearsum commented on src/services/releases.js:
```diff
@@ -1,10 +1,66 @@
+
+    parseReleases(response.data.items);
+
+    if (response.nextPageToken) {
```
I think this needs to be `response.data`.

mozbhearsum commented on src/utils/constants.js:
```diff
@@ -5,6 +5,12 @@ export const OBJECT_NAMES = {
 };
+export const GCS_CONFIG = {
+  NIGHTLY_HISTORY_BUCKET:
+    'https://www.googleapis.com/storage/v1/b/balrog-prod-nightly-history-v1/o',
```
This is going to differ depending on the environment we're deployed in. Can we read it from the environment? These are the correct URLs for local dev, so they can go in `.neutrinorc.js`.

mozbhearsum commented on src/utils/diff.js:
```diff
@@ -17,12 +33,8 @@ export default (diffProperties, objectOne, objectTwo) => {
-          typeof next === 'string' ? next.replace(NEW_LINES_REGEX, '') : next
-        }`,
+        prev: `${title(prop)}: ${formatValue(prev)}`,
+        next: `${title(prop)}: ${formatValue(next)}`,
```
If possible, can we avoid title casing propreties within the release data? It's jarring to have them differ between the full release data view and the diff (and the users who edit the blobs are accustomed to the specific property names).

helfi92 replied:
Agreed. We should probably remove it alltogether even from the rules view.

mozbhearsum commented on src/services/releases.js:
```diff
@@ -1,10 +1,66 @@
+  return getReleases(`${bucket}?prefix=${name}/&delimeter=/`);
+};
+
+const getRevisionData = link => axios.get(link);
```
Should the existing consumers of this just call `axios.get` instead, or is it bad form/architecture to do that directly in the views/components? I don't have super strong feelings here, it just stands out to see a function that does nothing except pass the arguments to another function.

helfi92 commented:
> The equivalent on the old UI (https://localhost:8010/releases/Devedition-58.0b10-build1) takes ~2s to load and render on my laptop.

> I also can't get the page to work correctly for releases with a lot of history like the one above - Firefox keeps throwing the slow script error until I decide to kill it.

I think this is because we are trying to display 1000 revisions using material UI components. The fix for that is to virtualize the list. This unacceptable performance hit is  the reason why I added the MVP label to https://github.com/mozilla-frontend-infra/balrog-ui/issues/104.

helfi92 commented:
i found a small UI glitch where the table header goes above the app bar when the table is big:

<img width="966" alt="Screen Shot 2019-08-07 at 9 03 34 PM" src="https://user-images.githubusercontent.com/3766511/62667715-e7469780-b956-11e9-9311-37998bff2837.png">

This will most probably be fixed in https://github.com/mozilla-frontend-infra/balrog-ui/issues/104.

helfi92 commented:
![Screen Shot 2019-08-09 at 11 19 08 AM](https://user-images.githubusercontent.com/3766511/62790029-334d2580-ba98-11e9-8679-275bb017c059.png)


A couple of changes:

1. Table is now virtualized
2. I changed the diff display to something similar to `git diff`. The diff display is virtualized to make sure long JSON don't slow down the page or take forever to render.
3. Diffing big JSONs with thousand of lines can take a long time. I extracted the diffing logic into a web worker to make sure it doesn't slow down the UI which also allows us to show a spinner while the diffing logic is running.

helfi92 commented:
> https://localhost:8010/releases/Devedition-56.0b1-build5

I'm not able to see that release for some reason even after searching for it in the releases view. All of the releases I'm seeing on that site seem to show only a couple of lines of diffs. Is it possible that we're diffing the wrong data in the new UI?

_Example: https://localhost:8010/releases/Devedition-56.0b3-build1_

![Screen Shot 2019-08-09 at 1 56 18 PM](https://user-images.githubusercontent.com/3766511/62798958-97c6af80-baad-11e9-9d4a-4e5feb52aec7.png)

mozbhearsum commented:
> I'm not able to see that release for some reason even after searching for it in the releases view.

That's a little surprising...but the production data dump is updated every night, so we're probably just working with slightly different datasets.

> All of the releases I'm seeing on that site seem to show only a couple of lines of diffs.

Yep, this is pretty normal. Each revision is typically a single locale being added to the Release, so just the few lines between data versions is expected. If you diff, eg: version 1000 vs 500 you'll see ~500 locales being added.

> Is it possible that we're diffing the wrong data in the new UI?

Possibly...I'm going to dig a little bit more.

mozbhearsum commented:
After some print debugging, it looks to me like `diffLines` is never returning when run on large strings. I tried to do some debugging with a simple script, and it looks like it's stuck in GC and one of the diff.js methods mostly:
![image](https://user-images.githubusercontent.com/49649/62801766-70271580-bab4-11e9-80d9-80991d23a60c.png)

helfi92 commented:
OK, I think I figured out why it's taking so long. What we need to do is deeply sort the objects first before comparing them.

helfi92 commented:
Filed https://github.com/mozilla-frontend-infra/balrog-ui/issues/145 to track the performance issue when viewing a release.

mozbhearsum commented:
The diffs still work, but I'm getting this exception whenever I try change the versions being compared, usually multiple times (possibly once per hunk of the diff?):
`Warning: Encountered two children with the same key, `+        },`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.`

I also saw an issue where there was weird looking hunks being shown, but I'm struggling to find a way to reproduce...

mozbhearsum commented:
The perf is much, much, much better! No more fan spinning forever :). A couple of other ideas for improvement (probably for later):
- Prefetch blobs onHover of the `View` button. Right now it takes long enough to be mildly annoying.
- Cache fetched blobs to improve successive diffs. We'd have to have a maximum size here to avoid blowing up memory usage, I suspect.

The other issue I saw is when I try comparing against a non-latest revision. For example, there's extra hunks at the top here:
![image](https://user-images.githubusercontent.com/49649/62946884-8af8d300-bdaf-11e9-955b-8f730bf70dc2.png)

It doesn't look like it happens with all releases, but I can always reproduce it with the CDM ones.

We can also see here that the revisions list is taking up more vertical space than necessary. Can we reduce that when there are few enough revisions that there's no scrolling?

This error is still present on first load of a revisions page too: `Warning: Failed prop type: Invalid prop `className` of type `string` supplied to `JsDiff`, expected `object`.`

helfi92 commented:
Added the requested changes. I had to re-virtualize jsDiff in case there are >1k lines of diff. If you compare revisions that are far apart, I was able to find revisions with >5k of line diffing. We would probably need to have the diffing happen in a worker so that we can maybe show a loader or something and not have the main thread hang but I think I'll do that in a follow-up. I'll just have to copy-paste from one of the earlier commits.

helfi92 commented:
Rebased.

mozbhearsum commented on src/components/Diff/index.jsx:
```diff
@@ -1,9 +1,9 @@
 import React, { useMemo } from 'react';
+import { object, oneOf } from 'prop-types';
```
I see this file is getting renamed, but is still only used for diffing Rules. Similarly, `JsDiff` is only used for diffing Releases. I've got no objection to them being generic, but I think they could use better names. It's not clear to me what the difference between them is based on the current names.
